src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java

```diff
@@ -341,6 +341,9 @@ private String writeRecordingToDestination(
         destination += ".jfr";
         try (InputStream stream = connection.getService().openStream(descriptor, false)) {
             fs.copy(stream, specificRecordingsPath.resolve(destination));
+        } catch (IOException ioe) {
```

Comment:
Assuming these are the correct docs, there are a few exceptions that `copy` could throw, including `IOException` and `SecurityException`. Not sure what other kinds of errors could cause an `IOException` though.

https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#copy-java.nio.file.Path-java.nio.file.Path-java.nio.file.CopyOption...-

from cryostat-core:
```
    public long copy(InputStream in, Path out, CopyOption... copyOptions) throws IOException {
        return Files.copy(in, out, copyOptions);
    }
```

Reply:
Yea, that would be the correct doc.

`FileAlreadyExistsException` and perhaps `DirectoryNotEmptyException` are specific IOException subclasses we could check for here, and just re-throw them as-is rather than wrapping them in an `EmptyRecordingException`.

Instead of trying to copy the stream and relying on an `IOException` to occur to inform us that the stream was in fact empty, could we use `stream.available() == 0`? I hope that `available()` on the recording stream is implemented in a way that helps us here.

If not, then perhaps checking for that `MBeanException` in the causal chain could be our indicator that we had an empty stream and not some other IOException? That's not ideal, but at least we can avoid misreporting other kinds of IOException as empty streams.

Reply:
I can try comparing the error message to "No recording data available" and see what happens

Reply:
Whoops, didn't see your reply until it refreshed. I'll try those suggestions

Reply:
Re-throwing `FileAlreadyExistsException` and `DirectoryNotEmptyException` works in this case.

`stream.available()` seems to always equal 0, just like the abstract implementation, and there doesn't seem to be an `MBeanException` in the causal chain:

```
if (stream.available() == 0) {
                fs.deleteIfExists(specificRecordingsPath.resolve(destination));
                throw new EmptyRecordingException("No recording data available");
            }
```
```
Error:    RecordingArchiveHelperTest.shouldSaveRecordingThatEndsWithJfr:475 » Execution ...
Error:    RecordingArchiveHelperTest.shouldSaveRecordingWithAlias:233 » Execution io.cry...
Error:    RecordingArchiveHelperTest.shouldSaveRecordingWithoutAlias:315 » Execution io....
Error:    RecordingArchiveHelperTest.shouldSaveRecordingWithoutServiceRef:392 » Execution
...
[ERROR] testWorkflow  Time elapsed: 4.008 s  <<< FAILURE!
java.lang.AssertionError: 
list-saved should have size 1 after recording save
Expected: <1>
     but: was <0>
	at itest.RecordingWorkflowIT.testWorkflow(RecordingWorkflowIT.java:171)
```

```
fs.copy(stream, specificRecordingsPath.resolve(destination));
} catch (IOException ioe) {
            if(ExceptionUtils.hasCause(ioe, MBeanException.class)) {
            fs.deleteIfExists(specificRecordingsPath.resolve(destination));
            throw new EmptyRecordingException(ioe);
            }
           throw ioe;
```
```
[ERROR] testSaveEmptyRecordingDoesNotArchiveRecordingFile  Time elapsed: 0.675 s  <<< ERROR!
java.util.concurrent.ExecutionException: io.vertx.ext.web.handler.impl.HttpStatusException: Internal Server Error
	at itest.TargetRecordingPatchIT.testSaveEmptyRecordingDoesNotArchiveRecordingFile(TargetRecordingPatchIT.java:119)
Caused by: io.vertx.ext.web.handler.impl.HttpStatusException: Internal Server Error
```

Reply:
Huh. That's unfortunate about `available()`.

The `MBeanException` I referenced came from #632 - strange that it isn't there in this case now.

Checking for the `"No recording data available"` cause string might be the best we can do then, unfortunately. I just worry that this message may not be consistent between target JVM versions.

Reply:
The stream implementation is here:

https://github.com/openjdk/jmc/blob/master/application/org.openjdk.jmc.rjmx.services.jfr/src/main/java/org/openjdk/jmc/rjmx/services/jfr/internal/JfrRecordingInputStreamV2.java

Looks like `available()` would work, but only after we have already begun reading some content from the stream. There doesn't appear to be any way to pre-fill the buffer to ensure it isn't empty without draining some of the initial bytes out of the stream.

Can we use a [BufferedInputStream](https://docs.oracle.com/javase/7/docs/api/java/io/BufferedInputStream.html) and `mark/reset` methods to pre-fill the buffer, determine if it's non-empty, and then reset to the start position to begin copying the contents to the output file?

Reply:
I tried to implement the following, except all the `shouldSaveRecording` unit tests are throwing an `EmptyRecordingException`:
```
        try (BufferedInputStream bufferedStream = new BufferedInputStream(connection.getService().openStream(descriptor, false))) {

            // Check if recording stream is non-empty
            int readLimit = 1000; // arbitrary number
            bufferedStream.mark(readLimit);

            if((bufferedStream.read() == -1) || (bufferedStream.available() == 0)) {
                fs.deleteIfExists(specificRecordingsPath.resolve(destination));
                throw new EmptyRecordingException();
            }

            bufferedStream.reset();

            fs.copy(bufferedStream, specificRecordingsPath.resolve(destination));
        }
```
Does the buffer get filled with data as soon as it's initialized from the `JfrRecordingInputStream`? What am I missing?

Reply:
That logic looks like what I would have expected. I don't believe the buffer is filled at initialization, but the `read()` call should do it.

Do the failing unit tests have an actual `InputStream` for the buffered stream to read from, or are they reading from mock streams?

Reply:
If they're reading from mock streams you can try replacing those with new ByteArrayInputStream("someText".getBytes()) - I think in this case the handler isn't going to validate what the stream content actually is, right? It just needs some nonempty stream of bytes to read from and attempt to copy.

Reply:
I tried the following with no luck:
```
        BufferedInputStream bufferedStream = new ByteArrayInputStream("someRecordingData".getBytes());
```
```
        InputStream recordingStream = new ByteArrayInputStream("someRecordingData".getBytes(Charset.forName("UTF-8")));
        BufferedInputStream bufferedStream = new BufferedInputStream(recordingStream);
```
```
        BufferedInputStream bufferedStream = Mockito.mock(BufferedInputStream.class);
        Mockito.when(bufferedStream.read()).thenReturn(1);
        Mockito.when(bufferedStream.available()).thenReturn(0);
```

When `String saveName =
                recordingArchiveHelper
                        .saveRecording(new ConnectionDescriptor(targetId), recordingName)
                        .get();` is called in the unit test, do all methods called inside `saveRecording()` become stubbed, including `writeRecordingToDestination`? How would I mock the methods inside`writeRecordingToDestination()` if the tests call `saveRecording()`?

Reply:
No, `writeRecordingToDestination` would not be a stubbed method unless you explicitly did that - normally, if your unit test calls `saveRecording`, then it will call the real `writeRecordingToDestination` (if it gets there).

To mock the things that are happening inside of `writeRecordingToDestination` you would need to set up the things that are mocked to behave however is needed to get the actual code under test to go down the execution pathway you're targeting. In this case, if you can call `saveRecording` and get it to the point that it's trying to check if the stream is empty, then you just need to do something like:

```java
IFlightRecorderService svc = Mockito.mock(IFlightRecorderService.class);
Mockito.when(connection.getService()).thenReturn(svc);
InputStream stream = new ByteArrayInputStream("someText".getBytes());
Mockito.when(svc.openStream(Mockito.any(IRecordingDescriptor.class), Mockito.anyBoolean())).thenReturn(stream);
```

before calling `helper.saveRecording()`. This way, when the code under test gets to the `try (InputStream stream = connection.getService().openStream(descriptor, false)) {` line, the stream it gets back is your `ByteArrayInputStream`. The helper code under test would be none the wiser and would go on to wrap that in the `BufferedInputStream` of its own, do the `mark/read/available/reset` stuff, etc.

Reply:
The hashcode for `stream` appears to be different at the end of the unit test. Could it be because it's a real, mutable object and not mocked?

Thanks for your help so far, we can continue next week.
```
void shouldSaveRecordingThatEndsWithJfr() throws Exception {
...
Mockito.verify(fs).copy(Mockito.eq(stream), Mockito.eq(destination));
...
```
```
[ERROR] Failures: 
[ERROR]   RecordingArchiveHelperTest.shouldSaveRecordingThatEndsWithJfr:488 
Argument(s) are different! Wanted:
fs.copy(
    java.io.BufferedInputStream@329aa2d1,
    Mock for Path, hashCode: 1482861352
);
-> at io.cryostat.core.sys.FileSystem.copy(FileSystem.java:91)
Actual invocations have different arguments:
fs.exists(
    Mock for Path, hashCode: 1968869542
);
-> at io.cryostat.recordings.RecordingArchiveHelper.writeRecordingToDestination(RecordingArchiveHelper.java:307)
fs.exists(
    Mock for Path, hashCode: 1482861352
);
-> at io.cryostat.recordings.RecordingArchiveHelper.writeRecordingToDestination(RecordingArchiveHelper.java:336)
fs.copy(
    java.io.BufferedInputStream@3c717ef2,
    Mock for Path, hashCode: 1482861352
);
-> at io.cryostat.recordings.RecordingArchiveHelper.writeRecordingToDestination(RecordingArchiveHelper.java:358)
```

Reply:
I think the problem there now is that the stream returned by the `openStream()` call is no longer the same stream instance passed to `fs.copy()` - since the implementation now wraps that stream in a `BufferedInputStream`. Rather than `Mockito.eq(stream)` it might be best to use `Mockito.isA(BufferedInputStream.class)`, although I don't know if there is a good way to verify that the buffered stream is wrapping around the original stream. I suppose Mockito constructor spies could be used for this.